x-pack/legacy/plugins/file_upload/public/util/file_parser.js

```diff
@@ -7,39 +7,121 @@
 import _ from 'lodash';
 import { geoJsonCleanAndValidate } from './geo_json_clean_and_validate';
 import { i18n } from '@kbn/i18n';
+import { PatternReader } from './pattern_reader';
 
-export async function readFile(file) {
-  const readPromise = new Promise((resolve, reject) => {
-    if (!file) {
-      reject(new Error(i18n.translate(
-        'xpack.fileUpload.fileParser.noFileProvided', {
+// In local testing, performance improvements leveled off around this size
+export const FILE_BUFFER = 1024000;
+
+const readSlice = (fileReader, file, start, stop) => {
+  const blob = file.slice(start, stop);
+  fileReader.readAsBinaryString(blob);
+};
+
+let prevFileReader;
+let prevPatternReader;
+export const fileHandler = async (
+  file, chunkHandler, cleanAndValidate, getFileParseActive,
+  fileReader = new FileReader(), fileBuffer = FILE_BUFFER
```

Comment:
remove fileBuffer from the signature.

Also, I would use named variables here, since there are so many.